### PR TITLE
Add REM docs sections and fill out the first section

### LINKED ISSUE #1451: REM: Write the first sub-section of the user guide
Per #1448, write the following documentation section:
  - How do I use REM?

**Considerations**:
- A rough outline for the documentation is [here](https://docs.google.com/document/d/1buO5PrO5sS02VXjcaYf37RuR0rF6xpyr4J9H1tI4vN4/edit?pli=1#heading=h.m66w3ecdwtem)

### PR DESCRIPTION
**Pull request by amirebrahimi**
## Description

Resolves #1451 by creating the REM docs section and adding content to the first part.

---

### License

- [X] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.
- [ ] I added unit tests for new code.
- [ ] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [ ] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [X] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [ ] Added myself / the copyright holder to the AUTHORS file

### CODE REVIEW

**Comment by amirebrahimi** [comment]
How does one fix an error like this:
```
broken    https://mitiq.readthedocs.io/en/latest/guide/rem.html - 404 Client Error: Not Found for url: https://mitiq.readthedocs.io/en/latest/guide/rem.html
```

The file doesn't exist because it is new and hasn't been published yet.

**Comment by natestemen** [comment]
I wish I had a better answer for you, but we typically just make a note that's why the linkcheck is failing and move forward. If this is ready otherwise, feel free to assign a reviewer!

**Comment by amirebrahimi** [comment]
I don't have permissions to assign a reviewer. It relies on someone from the mitiq assigning to themselves. That being said this is ready for review.

**Comment by andreamari** [comment]
Hi @amirebrahimi, no problem with the linkcheck error. It will self-fix once this PR is merged. 

This is not a proper review, but I would like to give two general comments before I forget about them:

1. Can we use more realistic errors e.g.  `p0=p1=0.01`? instead of the extreme case `p0=p1=1 ?` 
2. What about if we define the confusion matrix instead of its inverse? I think it's more intuitive. To get the inverse we can just use `np.linalg.inv()`.

P.S. Great docs!

File: docs/source/guide/rem-1-intro.myst [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+from mitiq import rem
+
+inverse_confusion_matrix = np.array(
+    [
+        [0, 0, 0, 1],
+        [0, 0, 1, 0],
+        [0, 1, 0, 0],
+        [1, 0, 0, 0],
+    ]
+)
```
Comment (natestemen) [inline docs/source/guide/rem-1-intro.myst:106]: Should probably explain where this comes from.

File: docs/source/guide/rem-1-intro.myst [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+noisy_value = raw_execute(circuit, noisy_executor, observable)
+
+ideal_executor = partial(noisy_readout_executor, p0=0, p1=0)
+ideal_value = raw_execute(circuit, ideal_executor, observable)
+print(f"Error without mitigation: {abs((ideal_value - noisy_value)/ideal_value) :.3}")
+```
+
+## Apply REM to raw readout results
+Readout-error mitigation can be easily implemented with the function
+{func}`.execute_with_rem()`.
```
Comment (natestemen) [inline docs/source/guide/rem-1-intro.myst:94]:
```suggestion
Readout-error mitigation can be easily applied with the function
{func}`.execute_with_rem()`.
```

File: docs/source/guide/rem-1-intro.myst [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+```
+
+Here we observe that the application of REM eliminates the readout error when compared
+to the unmitigated result.
+
+```{admonition} Note:
+It is necessary to supply the inverse confusion matrix to the REM technique.
+There are various approaches that can be used to generate the inverse 
+confusion matrix with some being more costly than others.
+```
```
Comment (natestemen) [inline docs/source/guide/rem-1-intro.myst:127]: I think this should work:
````suggestion
```{note} 
It is necessary to supply the inverse confusion matrix to the REM technique.
There are various approaches that can be used to generate the inverse 
confusion matrix with some being more costly than others.
```
````

I think it might be helpful if the link/sentence that follows this note is in the note as well since it's regarding the same topic.

**Review by natestemen — COMMENTED** [review]
Looking good! Some comments to add a touch more explanation in certain areas, but I like the demo.

Also nice job remembering to link to REM in the README.

File: docs/source/guide/rem.myst [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Readout-Error Mitigation
+
+Readout-Error Mitigation (REM) is an error mitigation technique in which
+inverted transition / confusion matrices are applied to the noisy results 
+of the readout.
```
[the remark this answers is not in the document]
Comment (natestemen) [reply docs/source/guide/rem.myst:5]: I think it might also be good to add that this technique goes by other names such as "measuremen error mitigation" in other places.

**Comment by amirebrahimi** [comment]
> Hi @amirebrahimi, no problem with the linkcheck error. It will self-fix once this PR is merged.
> 
> This is not a proper review, but I would like to give two general comments before I forget about them:
> 
> 1. Can we use more realistic errors e.g.  `p0=p1=0.01`? instead of the extreme case `p0=p1=1 ?`
> 2. What about if we define the confusion matrix instead of its inverse? I think it's more intuitive. To get the inverse we can just use `np.linalg.inv()`.
> 
> P.S. Great docs!

This was just out of habit as it was easy to specify the inverse confusion matrix explicitly. I replaced this with use of the helper function w/ explanation. I'll cover more about the confusion matrix in other sections.

File: docs/source/guide/rem.myst [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+<!--
+```{figure} ../img/rem_workflow.png
+---
+width: 700px
+name: rem-workflow-overview
+---
+Workflow of the REM technique in Mitiq, detailed in the [What happens when I use REM?](rem-4-low-level.myst) section.
+```
+-->
```
[the remark this answers is not in the document]
Comment (amirebrahimi) [reply docs/source/guide/rem.myst]: I've added this.

File: docs/source/guide/rem-1-intro.myst [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import mitiq
+
+mitiq.SUPPORTED_PROGRAM_TYPES.keys()
+```
+
+
+## Problem setup
+This example uses a Z0 + Z1 observable with a circuit that flips each qubit to 
+the |1> state. However, this is an incredibly noisy device where each qubit's
+readout is flipped. So, we mitigate these error results with REM.
```
[the remark this answers is not in the document]
Comment (amirebrahimi) [reply docs/source/guide/rem-1-intro.myst]: I've updated this.

File: docs/source/guide/rem-1-intro.myst [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+circuit = Circuit(X.on_each(*qreg), measure_each(*qreg))
+observable = Observable(PauliString("ZI"), PauliString("IZ"))
+
+print(circuit)
+```
+
+Next we define a simple noisy readout executor function which takes a 
+circuit as input, executes the circuit on a noisy simulator, and 
+returns the raw measurement results. See the [Executors](executors.myst) 
+section for more information on how to define more advanced executors.
```
[the remark this answers is not in the document]
Comment (amirebrahimi) [reply docs/source/guide/rem-1-intro.myst:48]: Added. Thanks for the suggestion.

File: docs/source/guide/rem-1-intro.myst (hunk 1 above)
Comment (amirebrahimi) [reply docs/source/guide/rem-1-intro.myst]: I've changed this code to use a helper method now.

File: docs/source/guide/rem.myst (hunk 4 above)
[the remark this answers is not in the document]
Comment (amirebrahimi) [reply docs/source/guide/rem.myst]: Added. Thanks.

**Comment by andreamari** [comment]
I also have a suggestion for the workflow diagram. 
image

I think it can be confusing to mention other techniques in the REM figure which is supposed to just explain what REM does. Note also that it is not obvious if additional techniques should come before REM (e.g. DDD) or after REM (CDR) or a partially before and partially after REM (e.g. ZNE).  

So I would suggest to simply remove the "other techniques" square from your figure and, optionally, write 
**1. Generate circuits (trivial for REM)**

Basically the REM figure should be kind of the opposite of the DDD figure (shown below). 

image


If we need to explain how to combine REM with other techniques, we can do it more specific subsections but I would keep the landing page as simple as possible.

**Review by natestemen — APPROVED** [review]
Nice work Amir!
